### PR TITLE
[Create Block] Adding a `--no-plugin` flag to scaffold only block files.

### PR DESCRIPTION
**Pull request by ryanwelcher**
## What?

This PR introduces the `--no-plugin` flag to allow users to use the package to scaffold only block files.

## Why?
Currently, there is no way to scaffold additional blocks once the plugin has been created. Being able to create subsequent blocks quickly would be very helpful.

## How?
When the `--no-plugin` flag is added, the script skips any plugin scaffolding and only outputs the files listed in the `blockTemplatesPath` ( just the block files ) into `folderName` as defined by the template with the default being `./src`. It should be noted that the script is not "aware" of its location. Calling `npx @wordpress/create-block block-name --no-plugin` will create a new block inside the `folderName`directory where it's called

The slug passed is used as the directory and block name any other flags not related to package.json items should function as expected for the block, i.e namespace. 

## Testing Instructions
1. From the root of `packages/create-block`, run `node index.js test ` to create a new plugin with a single block
2. `cd` into the `test` directory
3.  Run `node ../index.js my-test-block --no-plugin` to scaffold a new block
4. Confirm that the `src` directory has the existing block files and a new directory called `my-test-block` that contains new block files
5. Run `npm start` to confirm that the build works

### CODE REVIEW

**Comment by ryanwelcher** [comment]
This PR introduces a lot of conditionals and it would be a really good idea to start abstracting out some of the functionality to make it easier to work with.

File: packages/create-block/lib/scaffold.js
Comment (gziolo) [inline packages/create-block/lib/scaffold.js]:
```suggestion
```

File: packages/create-block/lib/scaffold.js
Comment (gziolo) [inline packages/create-block/lib/scaffold.js]:
```suggestion
```

File: packages/create-block/lib/output.js
Comment (gziolo) [inline packages/create-block/lib/output.js]: In theory, you could use the slug pattern also with non-PHP files. I'm fine with whatever you decide here.

File: packages/create-block/lib/init-block.js
Comment (gziolo) [inline packages/create-block/lib/init-block.js]: In case the block only flag, should we simply create a subdirectory named after `view.slug` with all templates? It would be the responsibility of the consumer to run the script in the correct place.

The other alternative would be to try to programmatically locate the `package.json` file and make everything relative to that so that would be very close to what's included:

```js
packageJsonPath ? join( packageJsonPath, view.folderName, view.slug, outputFile ) : join( view.slug, outputFile )
```

It still would need to remain as is for the case with the plugin's shell:

```js
join( view.folderName, outputFile )
```

File: packages/create-block/lib/output.js
Comment (gziolo) [reply packages/create-block/lib/output.js]: I see that being an issue in my testing:

Screenshot 2022-06-15 at 10 04 33

The message on the terminal doesn't align with the folder structure created:

```
To enter the directory type:

  $ cd my-block-1
```

That would be `cd src/my-block-1` in this case.

**Review by gziolo — COMMENTED** [review]
> This PR introduces the `--block-only` flag to allow users to use the package to scaffold only block files.

This is great. We really need this option to make it possible to scaffold more blocks for existing plugins ❤️ 

I left some comments from my testing. In general, this PR is nearly ready. It just requires some final decisions like:
- Where to exactly output the block folder?
- How to name the flag? It's probably too late to have `npx @wordpress/create-block` and `npx @wordpress/create-block-plugin` 😅 The only alternatives I can think of are `--no-plugin` or `--skip-plugin` which might be less clear.

File: packages/create-block/lib/output.js
Comment (ryanwelcher) [reply packages/create-block/lib/output.js]: I've updated the messaging here based on what is being created. I've also changed the prompt to move into the directory to only be displayed when a plugin is being generated. As scripts is in the root dir, there will be no need to cd into the block to run commands.

File: packages/create-block/lib/init-block.js
Comment (ryanwelcher) [reply packages/create-block/lib/init-block.js:73]: I quite like the idea of detecting the pacakage.json. This would also allow us to fire a warning to the user if there is not one found.

Comment (ryanwelcher) [reply packages/create-block/lib/init-block.js]: > In case the block only flag, should we simply create a subdirectory named after `view.slug` with all templates? It would be the responsibility of the consumer to run the script in the correct place.

This would require them to move into the `src` ( or whatever dir name they are using ) directory, correct? It would be simpler for those using custom src dir names as it's not impossible to pass the folder name to the CLI. 

Currently, the directory name is automatically picked up from the template used, but that doesn't always make sense depending on use-case.

Let's put it on the user to be in the correct place as an MVP and see what happens in the wild.

**Comment by ryanwelcher** [comment]
Based on a conversation with @gziolo in slack we think that changing the flag to `--no-plugin` makes more sense and is clearer in the context of running the full command i.e `npx @wordpress/create-block --no-plugin`

File: packages/create-block/lib/messaging.js
Comment (gziolo) [inline packages/create-block/lib/messaging.js]: We better avoid using `lodash`. @tyxla is working to remove its usage from the Gutenberg repository.

Comment (tyxla) [reply packages/create-block/lib/messaging.js]: Good point. Noting that this PR was created right when the Lodash removal work was starting, so it wasn't immediately clear we're getting rid of Lodash already.

Comment (gziolo) [reply packages/create-block/lib/messaging.js]: I included changes in https://github.com/WordPress/gutenberg/pull/43141 that set the minimum Node.js version to v14. As part of that, this PR removes all runs with Node.js 12, which should allow us to use optional chaining in this PR.

File: packages/create-block/CHANGELOG.md
Comment (gziolo) [inline packages/create-block/CHANGELOG.md:11]:
```suggestion
### New Feature

-   Add `--no-plugin` flag to allow scaffolding of a block in an existing plugin ([#41642](https://github.com/WordPress/gutenberg/pull/41642))
```

File: packages/create-block/lib/scaffold.js
Comment (gziolo) [inline packages/create-block/lib/scaffold.js]: I find the code more complex and harder to update with the new helper method introduced.
This could be as simple as:

```js
info(
    plugin
        ? `Creating a new WordPress plugin in the "${ slug }" directory.`
        : `Creating a new block in the "${ slug }" directory.`
);
```

I don't see anything wrong with branching the code in so many places for now.

File: packages/create-block/lib/init-block.js
Comment (gziolo) [inline packages/create-block/lib/init-block.js]: The code without negation in the conditional is easier to process:
```suggestion
	const outputFile = plugin
		? join( process.cwd(), slug, folderName, 'block.json' )
		: join( process.cwd(), slug, 'block.json' );
```

I see the same construct in other places. It's also probably an individual matter, so I will leave it up to you whether to leave it as is.

File: packages/create-block/lib/init-block.js
Comment (gziolo) [inline packages/create-block/lib/init-block.js]: I haven't tested it but for consistency we might want to include the full path here as well:
```suggestion
				: join( process.cwd(), view.folderName, outputFile );
```

**Review by gziolo — COMMENTED** [review]
Everything works great and I'm really excited about this new feature. I left my feedback for consideration. Let's also include an example with some details how it works in the README file.

File: packages/create-block/lib/messaging.js
Comment (ryanwelcher) [reply packages/create-block/lib/messaging.js]: I'll add the optional chaining back in and wait for that to be approved. I still need to address some of the other feedback as well.

File: packages/create-block/lib/scaffold.js
Comment (ryanwelcher) [reply packages/create-block/lib/scaffold.js]: Agreed. I was trying to simplify, but looks like I over-engineered 😄.

File: packages/create-block/lib/init-block.js
Comment (ryanwelcher) [reply packages/create-block/lib/init-block.js]: Updated.

File: packages/create-block/lib/init-block.js
Comment (ryanwelcher) [reply packages/create-block/lib/init-block.js]: Updated.

**Comment by ryanwelcher** [comment]
I found an issue while testing where if a `--template` was passed that didn't support the `blockTemplatesPath` property an `assets` directory was created containing the actual template files.

As a result, I have added some logic to check to see if there are block template files and if not, error out with a message.

This can be tested by using the following command from the root of the create-block package.

1. Template with support: `node ../index.js my-test-block ---template=@ryanwelcher/dynamic-block-template --no-plugin`
2. Template without support:`node ../index.js my-test-block --template=@ryanwelcher/multiple-blocks-template --no-plugin`

File: packages/create-block/lib/init-block.js
Comment (ryanwelcher) [reply packages/create-block/lib/init-block.js]: I had to revert this as it seemed to be creating a `Users/path/to/where/I/called/it` directory in the `src` when scaffolding.

Comment (gziolo) [reply packages/create-block/lib/init-block.js]: Good 👍

**Comment by gziolo** [comment]
I can confirm that the newly added check for block templates works as expected:

Screenshot 2022-08-12 at 09 46 31

It makes me wonder whether we should bail out early when loading the config for the template. In effect we would have two checks:
- Show error when there is no `blockTemplatesPath` provided together with `--no-plugin` flag: `The template selected doesn't support `--no-plugin` option.
- If both `--no-plugin` and the `blockTemplatesPath` was provided and there were no block files loaded, we could just show the message that `No block files found in the template. The template "XYZ" is incorrectly configured`.

**Review by gziolo — APPROVED** [review]
Excellent work. I think it's ready feature-wise. There is only one remaining task left for the `lodash` support, which isn't a blocker but nice to have.

The other comment about handling edge cases is my just musing that you can ignore in the first iteration 😄

**Comment by gziolo** [comment]
https://github.com/WordPress/gutenberg/blob/9a4b95e37da8b4c4b271f510895690a4ea1fa780/packages/create-block/lib/prompts.js#L11

here we could make it more universal for the time being so it fits also the case when user passes `--no-plugin` flag:

`The block slug used for identification (also the output folder name):`

In the long run, if we were to support multiple blocks scaffolding in one go, we might want to have two distinct slug prompts - one for the block and the other one for the plugin. I'm not sure about that yet, but I wanted to flag it.

**Comment by ryanwelcher** [comment]
> https://github.com/WordPress/gutenberg/blob/9a4b95e37da8b4c4b271f510895690a4ea1fa780/packages/create-block/lib/prompts.js#L11
> 
> here we could make it more universal for the time being so it fits also the case when user passes `--no-plugin` flag:
> 
> `The block slug used for identification (also the output folder name):`
> 

I'll make this update before merging.

> In the long run, if we were to support multiple blocks scaffolding in one go, we might want to have two distinct slug prompts - one for the block and the other one for the plugin. I'm not sure about that yet, but I wanted to flag it.

I LOVE this idea. It's been something that always bugged me that the plugin and the block had the same name. Being able to pass different names for the plugin would be really handy. I'll open a new issue for it.

**Comment by ryanwelcher** [comment]
related https://github.com/WordPress/gutenberg/issues/43235

**Pull request merged**
